File: backend/Litres.Infrastructure/Outbox/OutboxProcessor.cs
using System.Text.Json;
using Litres.Application.Protos;
using MassTransit;
using Microsoft.EntityFrameworkCore;

namespace Litres.Infrastructure.Outbox;

public sealed class OutboxProcessor(ApplicationDbContext dbContext, IPublishEndpoint publishEndpoint)
{
    private const int BatchSize = 10;

    public async Task<int> Execute(CancellationToken ctx = default)
    {
        var outboxMessages = await dbContext.OutboxMessages
            .Where(m => m.ProcessedOn == null)
            .OrderBy(m => m.OccuredOn)
            .Take(BatchSize)
            .ToListAsync(ctx);
        
        foreach (var outboxMessage in outboxMessages)
        {
            try
            {
                var messageType = typeof(OutboxMessage).Assembly.GetType(outboxMessage.Type)!;
                var deserializedMessage = JsonSerializer.Deserialize(outboxMessage.Content, messageType)!;

                await publishEndpoint.Publish(deserializedMessage, messageType, ctx);

                outboxMessage.ProcessedOn = DateTime.UtcNow;
                dbContext.OutboxMessages.Update(outboxMessage);
            }
            catch (Exception ex)
            {
                outboxMessage.ProcessedOn = DateTime.UtcNow;
                outboxMessage.Error = ex.ToString();
                dbContext.OutboxMessages.Update(outboxMessage);
            }
        }
        
        await dbContext.SaveChangesAsync(ctx);
        return outboxMessages.Count;
    }
}

File: backend/Litres.Payment/src/Outbox/OutboxProcessor.ts
import { DataSource } from 'typeorm';
import { OutboxMessage } from '../Models/OutboxMessage';
import { IPublishEndpoint } from './IPublishEndpoint';

export class OutboxProcessor {
    private readonly publishEndpoint: IPublishEndpoint;
    private readonly batchSize = 10;
    private readonly dataSource: DataSource;

    constructor(dataSource: DataSource, publishEndpoint: IPublishEndpoint) {
        this.dataSource = dataSource;
        this.publishEndpoint = publishEndpoint;
    }

    public async execute(ctx: any = {}): Promise<number> {
        const outboxMessageRepository = this.dataSource.getRepository(OutboxMessage);

        const outboxMessages = await outboxMessageRepository
            .createQueryBuilder('outboxMessage')
            .where('outboxMessage.processedOnUtc IS NULL')
            .orderBy('outboxMessage.occuredOnUtc', 'ASC')
            .take(this.batchSize)
            .getMany();

        for (const outboxMessage of outboxMessages) {
            try {
                const messageType = require(outboxMessage.type).default;
                const deserializedMessage = JSON.parse(outboxMessage.content);

                await this.publishEndpoint.publish(deserializedMessage, messageType, ctx);

                outboxMessage.processedOn = new Date();
                await outboxMessageRepository.save(outboxMessage);
            } catch (ex) {
                outboxMessage.processedOn = new Date();
                outboxMessage.error = ex.toString();
                await outboxMessageRepository.save(outboxMessage);
            }
        }

        return outboxMessages.length;
    }
}